rlclientlib/federation/vw_local_joiner.cc

```diff
@@ -0,0 +1,191 @@
+  _joiner = VW::make_unique<example_joiner>(_joiner_workspace.get());
+  _joiner->set_problem_type_config(messages::flatbuff::v2::ProblemType_CB);
+  _joiner->set_learning_mode_config(messages::flatbuff::v2::LearningModeType_Online);
+  _joiner->set_reward_function(messages::flatbuff::v2::RewardFunctionType_Earliest);
+  _joiner->set_default_reward(0.f);
+}
+
+int vw_local_joiner::add_events(const std::vector<buffer>& events, api_status* status)
+{
+  for (const auto& data : events)
+  {
+    logger::preamble pre;
+    pre.read_from_bytes(data->preamble_begin(), pre.size());
+    if (pre.msg_type != logger::message_type::fb_generic_event_collection)
+    {
+      RETURN_ERROR_LS(_trace_logger, status, invalid_argument)
+          << " Message type " << pre.msg_type << " cannot be handled.";
+    }
+
+    auto res = messages::flatbuff::v2::GetEventBatch(data->body_begin());
+    flatbuffers::Verifier verifier(data->body_begin(), data->body_filled_size());
+    if (!res->Verify(verifier))
+    { RETURN_ERROR_LS(_trace_logger, status, invalid_argument) << "verify failed for fb_generic_event_collection"; }
+
+    if (res->metadata()->content_encoding()->str() != "IDENTITY")
+    { RETURN_ERROR_LS(_trace_logger, status, invalid_argument) << "Can only handle IDENTITY encoding"; }
+
+    for (auto message : *res->events())
+    {
+      flatbuffers::FlatBufferBuilder fbb;
+      // TODO: real timestamp
+      messages::flatbuff::v2::TimeStamp ts(2020, 3, 1, 10, 20, 30, 0);
+      const auto* event_payload = message->payload();
+      auto vec = fbb.CreateVector(event_payload->data(), event_payload->size());
+      auto fb = messages::flatbuff::v2::CreateJoinedEvent(fbb, vec, &ts);
+      fbb.Finish(fb);
+      auto detached_buffer = fbb.Release();
+
+      const auto* je = flatbuffers::GetRoot<messages::flatbuff::v2::JoinedEvent>(detached_buffer.data());
+      _joiner->process_event(*je);
+    }
+  }
+
+  return error_code::success;
+}
+
+int vw_local_joiner::invoke_join(std::unique_ptr<i_joined_log_batch>& batch_out, api_status* status)
+{
+  std::stringstream ss;
+  ss << "Joining " << _joiner->events_in_queue() << " events." << std::endl;
+  TRACE_INFO(_trace_logger, ss.str());
+
+  std::vector<VW::example*> examples;
+  std::unique_ptr<vw_joined_log_batch> joined_batch = VW::make_unique<vw_joined_log_batch>(_joiner_workspace);
+
+  while (_joiner->processing_batch())
+  {
+    examples.push_back(VW::new_unused_example(*_joiner_workspace));
+
+    // False means there was a problem and we should try reading the next one.
+    if (!_joiner->process_joined(examples))
+    {
+      assert(examples.size() == 1);
+      auto* ex_to_return = examples.back();
+      examples.pop_back();
+      VW::finish_example(*_joiner_workspace, *ex_to_return);
+      continue;
+    }
+    VW::setup_examples(*_joiner_workspace, examples);
+
+    // We must remove the trailing newline example.
+    auto* newline_ex = examples.back();
+    examples.pop_back();
```

Comment:
Aren't we concatenating multiple multi_exs into single one here?

Reply:
This was from Jack's code I believe, maybe he can help clarify here.

Reply:
Yeah it seems like it could be alexey.  But I don't remember what's going on on the other end here. I don't think I am using the driver so newline examples are not relevant. I would not take the code as it was as correct since it was a very quick prototype

Reply:
I see. @byronxu99, maybe we can add test that is checking this? like adding to interactions with different ids and validating that we have 2 multi_ex on other end (now it looks like we will get single one).